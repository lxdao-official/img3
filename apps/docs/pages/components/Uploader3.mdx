---
title: Uploader3
description: Uploader3 is a React-based Web3 image upload component that supports multiple image uploads, image cropping, and uploading images to Web3 Storage providers (like IPFS). There are two ways for uploading, by using a backend API or the Uploader3 Connector.
---

# Uploader3

Uploader3 is a React-based Web3 image upload component that supports multiple image uploads, image cropping, and uploading images to Web3 Storage providers (like IPFS). There are two ways for uploading, by using a backend API or the [Uploader3 Connector](./uploader3-connector).

**Features**

- Supports cropping
- Supports uploading to web3 service providers, such as NFT.storage

## Usage

```js copy
import { Uploader3 } from '@lxdao/uploader3';
```

## Props

| Prop         | Type                 | Description                                            | Default                          |
| ------------ | -------------------- | ------------------------------------------------------ | -------------------------------- |
| accept       | `string`             | image accept file type                                 | `['.png','.jpeg','.jpg','.gif']` |
| multiple     | `boolean`            | multiple image upload                                  | `false`                          |
| api          | `string`             | endpoint upload api url                                | `''`                             |
| connector    | `object`             | create by [uploader3-connector](./uploader3-connector) | -                                |
| crop         | `[Crop]` / `boolean` | crop config, set `false` disabled crop                 | `true`                           |
| onChange     | `function`           | callback when files selected                           | -                                |
| onUpload     | `function`           | callback when file uploading                           | -                                |
| onComplete   | `function`           | callback when file uploaded                            | -                                |
| onCropEnd    | `function`           | callback when crop end                                 | -                                |
| onCropCancel | `function`           | callback when crop cancel                              | -                                |

<Callout type="info">
  `api` and `connector` are mutually exclusive, if both are provided, `api` will be used. must be provided one of them.
</Callout>

## Examples

import uploader3Examples from '@/examples/uploader3/';

### Basic

Uploader3 children is a button, you can use any component as children.

<Playground files={uploader3Examples.basic} />

### Single file upload

Upload a single image, crop it to a 1:1 ratio, and preview it.

<Playground files={uploader3Examples.single} />
