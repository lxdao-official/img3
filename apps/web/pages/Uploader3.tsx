import * as React from 'react';
import { Uploader3 } from 'uploader3';

export default function App() {
  return (
    <div style={{ padding: 40 }}>
      <Uploader3>
        <div
          style={{
            borderRadius: 5,
            padding: '20px',
            display: 'inline-block',
            boxShadow: '0 0 2px 0 rgba(0, 0, 0, 0.8) inset',
            backgroundColor: '#0987ff',
            color: '#fff',
          }}
        >
          Drop files or Click to select files
        </div>
      </Uploader3>
    </div>
  );
}
