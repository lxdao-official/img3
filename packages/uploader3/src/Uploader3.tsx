import React, { useCallback, useMemo, useState } from 'react';
import { useDropzone } from 'react-dropzone';
import styled from 'styled-components';

import { file2base64 } from './file2base64';
import { UploaderCrop, UploaderCroppProps } from './UploaderCrop';
import { useFiles } from './useEditFile';

import type { Uploader3Connector } from '@lxdao/uploader3-connector';
import type { CroppedFile, SelectedFile, SelectedFiles, Uploader3Props } from './types';

const Wrapper = styled.div`
  position: relative;
  display: inline-block;
  cursor: pointer;
`;

const createCroppedFile = (
  file: SelectedFile,
  croppAttributes: Pick<CroppedFile, 'imageData' | 'thumbData' | 'crop'>
): CroppedFile => {
  return { ...file, ...croppAttributes, status: 'cropped' };
};

const defaultCropOptions = {
  size: { width: '500px', height: '400px' },
  aspectRatio: 1,
};

export const Uploader3: React.FC<React.PropsWithChildren<Uploader3Props>> = ({
  children,
  className,
  style,
  multiple = false,
  accept = ['.jpg', '.jpeg', '.png', '.gif'],
  api,
  crop = true,
  headers,
  connector,
  responseFormat,
  onComplete,
  onUpload,
  onCropEnd,
  onCropCancel,
  onChange,
}: Uploader3Props) => {
  const {
    currentFile,
    currentIndex,
    setCurrentFile,
    setCurrentIndex,
    setFiles,
    files: selectedFiles,
  } = useFiles<any[]>([]);

  let cropOptions: any = crop;

  if (cropOptions === true) {
    cropOptions = defaultCropOptions;
  }

  const enableCrop = useMemo(() => !!cropOptions || false, [cropOptions]);
  const [showCrop, setShowCrop] = useState(false);

  const triggerComplete = (file: any) => {
    onComplete?.(file);
  };

  const doUpload = (files: any) => {
    if (!api && !connector) {
      throw new Error('Either api or connector must be provided');
    }

    Promise.all(
      files.map(async (curFile: any) => {
        curFile = { ...curFile };
        curFile.status = 'uploading';
        if (!curFile.imageData) {
          curFile.imageData = await file2base64(curFile.file);
        }
        onUpload?.(curFile);
        const image: Uploader3Connector.PostImageFile = { data: curFile.imageData, type: curFile.type };
        if (api) {
          return fetch(api, {
            method: 'POST',
            headers: {
              'Content-Type': 'application/json',
              ...headers,
            },
            body: JSON.stringify(image),
          })
            .then(async (res) => {
              if (res.ok) {
                let responseData = await res.json();

                if ('function' === typeof responseFormat) {
                  responseData = responseFormat(responseData);
                }

                const { url } = responseData;

                curFile = { ...curFile, status: 'done', url };
              } else {
                const { message } = await res.json();
                curFile = { ...curFile, status: 'error', message };
              }
              triggerComplete(curFile);
            })
            .catch(() => {
              curFile = { ...curFile, status: 'error' };
              triggerComplete(curFile);
            });
        } else if (connector) {
          return connector
            .postImage(image)
            .then((result: { url: string; cid: string }) => {
              const { url, cid } = result;
              curFile = { ...curFile, status: 'done', url, ipfs: 'ipfs://' + cid };
              triggerComplete(curFile);
            })
            .catch(() => {
              curFile = { ...curFile, status: 'error' };
              triggerComplete(curFile);
            });
        }
      })
    ).then(() => {
      // all done
    });
  };

  const afterCroppAction = useCallback(
    (file?: CroppedFile) => {
      if (currentIndex === selectedFiles.length - 1) {
        setShowCrop(false);
      } else {
        let nextIndex = currentIndex + 1;
        setCurrentIndex(nextIndex);
      }

      if (file) {
        doUpload([file]);
      }
    },
    [currentIndex, selectedFiles]
  );

  const onDrop = useCallback(async (acceptedFiles: File[]) => {
    const files: SelectedFiles = acceptedFiles.map((file) => {
      const { name, type } = file;
      return { name, type, file, previewUrl: URL.createObjectURL(file), status: 'none' };
    });

    setFiles(files);
    onChange?.(files);

    if (enableCrop) {
      setShowCrop(true);
    } else {
      doUpload(files);
    }
  }, []);

  const { getRootProps, getInputProps } = useDropzone({
    accept: { 'image/*': accept! },
    multiple,
    onDrop,
  });

  const handleCancel: Required<UploaderCroppProps>['onCancel'] = useCallback(() => {
    currentFile.status = 'cancel';
    onCropCancel?.({ ...currentFile });
    afterCroppAction();
  }, [currentFile]);

  const handleConfirm: Required<UploaderCroppProps>['onConfirm'] = useCallback(
    (params) => {
      const { imageData, thumbData, cropData } = params;
      const croppedFile = createCroppedFile(currentFile, {
        imageData,
        thumbData,
        crop: cropData,
      });
      setCurrentFile(croppedFile);
      onCropEnd?.(croppedFile);
      afterCroppAction(croppedFile);
    },
    [currentFile]
  );

  return (
    <>
      <Wrapper style={style} {...getRootProps({ className })}>
        {children}
        <input {...getInputProps()} />
      </Wrapper>
      {currentFile && cropOptions ? (
        <UploaderCrop
          size={cropOptions.size!}
          aspectRatio={cropOptions.aspectRatio!}
          show={showCrop}
          fileType={currentFile.type}
          fileUrl={currentFile.previewUrl!}
          onCancel={handleCancel}
          onConfirm={handleConfirm}
        />
      ) : null}
    </>
  );
};
