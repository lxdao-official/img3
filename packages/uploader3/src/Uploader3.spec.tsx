import * as React from 'react';

import { act, render, screen, fireEvent } from '@testing-library/react';
import { Uploader3 } from './Uploader3';

describe('src/Uploader3.tsx', function () {
  it('should render Uploader3', function () {
    const { container } = render(
      <Uploader3
        api={'/api/upload/file?name=your-name'}
        headers={{
          'x-token': 'abcd',
        }}
        multiple={false}
        crop={true} // use default crop options
        onChange={(files) => {
          {
            /*setFile(files[0]);*/
          }
        }}
        onUpload={(file) => {
          // setFile(file);
        }}
        onComplete={(file) => {
          // setFile(file);
        }}
        onCropCancel={(file) => {
          // setFile(null);
        }}
        onCropEnd={(file) => {
          // setFile(file);
        }}
      >
        <button>Upload</button>
      </Uploader3>
    );
    expect(container).toMatchSnapshot();
  });
});
