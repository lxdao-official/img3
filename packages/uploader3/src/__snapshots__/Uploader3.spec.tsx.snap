// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/Uploader3.tsx should render Uploader3 1`] = `
<div>
  <div
    class="sc-pGaPU dXHKSN"
    role="presentation"
    tabindex="0"
  >
    <button>
      Upload
    </button>
    <input
      accept="image/jpeg,.jpg,.jpeg,image/png,.png,image/gif,.gif,image/svg+xml,.svg"
      style="display: none;"
      tabindex="-1"
      type="file"
    />
  </div>
</div>
`;
